native/src/main/java/io/ballerina/lib/data/csvdata/csv/CsvTraversal.java

```diff
@@ -0,0 +1,849 @@
+        }
+
+        private static boolean ignoreRow(int index, Object skipLinesConfig) {
+            long[] skipLines = CsvUtils.getSkipDataRows(skipLinesConfig);
+            for (long skipLine: skipLines) {
+                if (skipLine == index) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        public Object initStatesForCsvRowWithMappingAsExpectedType(Object csvElement, Type expectedType) {
+            expectedType = TypeUtils.getReferredType(expectedType);
+            switch (expectedType.getTag()) {
+                case TypeTags.RECORD_TYPE_TAG:
+                    RecordType recordType = (RecordType) expectedType;
+                    this.fieldHierarchy = new HashMap<>(recordType.getFields());
+                    fields = new HashSet<>(recordType.getFields().keySet());
+                    this.updatedRecordFieldNames = CsvUtils.processNameAnnotationsAndBuildCustomFieldMap(
+                            recordType, fieldHierarchy);
+                    this.headerFieldHierarchy = new HashMap<>(recordType.getFields());
+                    this.restType = recordType.getRestFieldType();
+                    currentCsvNode = ValueCreator.createRecordValue(recordType.getPackage(), recordType.getName());
+                    traverseCsvRowWithMappingAsExpectedType(csvElement, expectedType, false);
+                    break;
+                case TypeTags.MAP_TAG:
+                    MapType mapType = (MapType) expectedType;
+                    currentCsvNode = ValueCreator.createMapValue(mapType);
+                    traverseCsvRowWithMappingAsExpectedType(csvElement, expectedType, true);
+                    break;
+                default:
+                    throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE, expectedType);
+            }
+            return currentCsvNode;
+        }
+
+        public Object initStatesForCsvRowWithListAsExpectedType(Object csvElement, Type expectedType) {
+            expectedType = TypeUtils.getReferredType(expectedType);
+            switch (expectedType.getTag()) {
+                case TypeTags.ARRAY_TAG:
+                    ArrayType arrayType = (ArrayType) expectedType;
+                    currentCsvNode = ValueCreator.createArrayValue(arrayType);
+                    traverseCsvRowWithListAsExpectedType(csvElement, arrayType);
+                    break;
+                case TypeTags.TUPLE_TAG:
+                    TupleType tupleType = (TupleType) expectedType;
+                    this.restType = tupleType.getRestType();
+                    currentCsvNode = ValueCreator.createTupleValue(tupleType);
+                    traverseCsvRowWithListAsExpectedType(csvElement, tupleType);
+                    break;
+                default:
+                    throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE, expectedType);
+            }
+            return currentCsvNode;
```

Comment:
I think we can use enhanced switch here as well

```suggestion
            currentCsvNode = switch (expectedType.getTag()) {
                case TypeTags.ARRAY_TAG -> ValueCreator.createArrayValue((ArrayType) expectedType);
                case TypeTags.TUPLE_TAG -> {
                    TupleType tupleType = (TupleType) expectedType;
                    this.restType = tupleType.getRestType();
                    yield ValueCreator.createTupleValue(tupleType);
                }
                default -> throw DiagnosticLog.error(DiagnosticErrorCode.INVALID_TYPE, expectedType);
            };
            traverseCsvRowWithListAsExpectedType(csvElement, expectedType);
            return currentCsvNode;
```